storage/src/adb/current.rs

```diff
@@ -331,26 +353,325 @@ impl<
+        let Some(op_count) = op_count else {
+            debug!("verification failed, invalid proof size");
+            return Ok(false);
+        };
+        let end_loc = start_loc + ops.len() as u64 - 1;
+        if end_loc >= op_count {
+            debug!(
+                loc = end_loc,
+                op_count, "proof verification failed, invalid range"
+            );
+            return Ok(false);
+        }
+
+        let start_pos = leaf_num_to_pos(start_loc);
+        let end_pos = leaf_num_to_pos(end_loc);
+
+        let mut hasher = Standard::new(hasher);
+        let digests = ops
+            .iter()
+            .map(|op| Any::<E, _, _, _, T>::op_digest(&mut hasher, op))
+            .collect::<Vec<_>>();
+
+        let chunk_vec = chunks.iter().map(|c| c.as_ref()).collect::<Vec<_>>();
+        let mut verifier = GraftingVerifier::new(
+            hasher.inner(),
+            Self::grafting_height(),
+            start_loc / Bitmap::<H, N>::CHUNK_SIZE_BITS,
+            chunk_vec,
+        );
+
+        if op_count % Bitmap::<H, N>::CHUNK_SIZE_BITS == 0 {
+            return proof
+                .verify_range_inclusion(&mut verifier, digests, start_pos, end_pos, root_digest)
+                .await
+                .map_err(Error::MmrError);
+        }
+
+        // The proof must contain the partial chunk digest as its last hash.
+        if proof.digests.is_empty() {
+            debug!("proof has no digests");
+            return Ok(false);
+        }
+        let mut proof = proof.clone();
+        let last_chunk_digest = proof.digests.pop().unwrap();
+
+        // Reconstruct the MMR root.
+        let mmr_root = match proof
+            .reconstruct_root(&mut verifier, digests, start_pos, end_pos)
+            .await
+        {
+            Ok(root) => root,
+            Err(MissingDigests) => {
+                debug!("Not enough digests in proof to reconstruct root");
+                return Ok(false);
+            }
+            Err(ExtraDigests) => {
+                debug!("Not all digests in proof were used to reconstruct root");
+                return Ok(false);
+            }
+            Err(e) => return Err(Error::MmrError(e)),
+        };
+
+        let next_bit = op_count % Bitmap::<H, N>::CHUNK_SIZE_BITS;
+        let reconstructed_root = Bitmap::<H, N>::partial_chunk_root(
+            hasher.inner(),
+            &mmr_root,
+            next_bit,
+            &last_chunk_digest,
+        );
+
+        Ok(reconstructed_root == *root_digest)
+    }
+
+    /// Generate and return a proof of the current value of `key`, along with the other
+    /// [KeyValueProofInfo] required to verify the proof. Returns KeyNotFound error if the key is
+    /// not currently assigned any value.
+    pub async fn key_value_proof(
+        &self,
+        hasher: &mut H,
+        key: K,
+    ) -> Result<(Proof<H>, KeyValueProofInfo<K, V, N>), Error> {
+        let op = self.any.get_with_loc(&key).await?;
```

Comment:
nit: recommend destructuring here to avoid opaque tuple access:

```
let Some((value, pos)) = self.any.get_with_loc(&key).await? else {
  return Err(Error::KeyNotFound());
} 
```

Reply:
fixed